extensions/amp-live-list/0.1/poller.js

```diff
@@ -0,0 +1,186 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {exponentialBackoff} from '../../../src/exponential-backoff';
+
+
+/**
+ * Polling state of a Poller instance.
+ * @enum {string}
+ */
+export const PollerState = {
+  /**
+   * The poller is running at a the set interval.
+   */
+  RUNNING: 'running',
+
+  /**
+   * The poller is running but is doing exponential backoff (using
+   * the original interval value as base) every tick.
+   */
+  BACKEDOFF: 'backedoff',
+
+  /**
+   * The poller is not running and not executing any work.
+   */
+  STOPPED: 'stopped',
+};
+
+
+/**
+ * Poller with backoff functionality.
+ */
+export class Poller {
+
+  constructor(win, wait, work) {
+    /** @private @const {!Window} */
+    this.win = win;
+
+    /** @private {number} */
+    this.wait_ = wait;
+
+    /** @private {function(): !Promise} */
+    this.work_ = work;
+
+    /** @private {?number} */
+    this.lastTimeoutId_ = null;
+
+    /** @private {?number} */
+    this.lastBackoffTimeoutId_ = null;
+
+    /** @private {function(): number} */
+    this.backoffPoller_ = this.getBackoffPoller_();
+
+    /** @private {!PollerState} */
+    this.state_ = PollerState.STOPPED;
+
+    /**
+     * Mostly for testing purposes.
+     * @private {!Promise}
+     */
+    this.lastWorkPromise_ = null;
+  }
+
+  /**
+   * @return {function(): number}
+   * @private
+   */
+  getBackoffPoller_() {
+    return exponentialBackoff();
+  }
+
+  /**
+   * @return {!PollerState}
+   */
+  getState() {
+    return this.state_;
+  }
+
+  /**
+   * Initalize any work needed to start polling.
+   */
+  start() {
+    if (this.lastTimeoutId_) {
+      return;
+    }
+    if (this.lastBackoffTimeoutId_) {
+      this.win.clearTimeout(this.lastBackoffTimeoutId_);
+    }
+
+    // we reset the backoff poller
+    this.backoffPoller_ = this.getBackoffPoller_();
+    this.state_ = PollerState.RUNNING;
+    this.poll_();
+  }
+
+  /**
+   * Initialize backoff function and set correct state.
+   */
+  backoff() {
+    // if theres already backoff work queued, then cancel this invocation.
+    if (this.lastBackoffTimeoutId_) {
+      return;
+    }
+    this.stop();
+
+    this.state_ = PollerState.BACKEDOFF;
+    this.backoff_();
+  }
+
+  /**
+   * Clear timers and set correct stop state.
+   */
+  stop() {
+    this.state_ = PollerState.STOPPED;
+    if (this.lastTimeoutId_) {
+      this.win.clearTimeout(this.lastTimeoutId_);
+      this.lastTimeoutId_ = null;
+    }
+    if (this.lastBackoffTimeoutId_) {
+      this.win.clearTimeout(this.lastBackoffTimeoutId_);
+      this.lastBackoffTimeoutId_ = null;
+    }
+  }
+
+  /**
+   * @private
+   */
+  backoff_() {
+    if (this.state_ != PollerState.BACKEDOFF) {
+      return;
+    }
+    this.lastBackoffTimeoutId_ = this.backoffPoller_(() => {
+      this.lastWorkPromise_ = this.work_().then(this.backoff_.bind(this))
```

Comment:
@dvoytenko yes, it actually does read illogically, whats worse is we actually cant control that since the recursive call inside `exponential-backoff` is outside of our control so ive removed the `then` handler.